enhancements/20240709-longhorn-v2-data-engine-v1.4.0.md

```diff
@@ -0,0 +1,174 @@
+
+### Implementation Overview
+
+#### Enabling the V2 Data Engine
+
+When the user enables the V2 Data Engine, we need to do the following:
+1. On each node:
+    - Load the relevant kernel modules: `modprobe vfio_pci ; modprobe uio_pci_generic ; modprobe nvme_tcp` (note: this is the set from [os2#98], vs. the [Longhorn Quick Start] which doesn't mention `vfio_pci`.  TODO: check this)
+    - Allocate 1024 hugepages: `echo 1024 > /sys/kernel/mm/hugepages/hugepages-2048kB/nr_hugepages`
+    - The above can be done at runtime (saves rebooting each node immediately), but these settings need to persist after reboot, so must be set in `/oem/93_spdk.yaml` as in [os2#98]
+2. Set Longhorn's `lhs/v2-data-engine` value to `true`.
+
+There are some kinks:
+
+- Ideally, it would be possible to make the above changes to all nodes without requring a reboot, but hugepage allocation isn't guaranteed to succeed if the system doesn't have enough available unfragmented memory. So we need to first try to allocate hugepages on each node, and if any fail (i.e. if `cat /sys/kernel/mm/hugepages/hugepages-2048kB/nr_hugepages` would print a number less than 1024), report that to the user and request the user reboot the affected node(s), so that hugepages can be allocated at boot time from `/oem/93_spdk.yaml`.
+- We actually don't want to always create `/oem/93_spdk.yaml` on every node, because this will allocate 1024 huge pages (i.e. it will eat 2GiB RAM), even if the V2 Data Engine is not enabled.
+    - We should be able to solve this by teaching node-manager to create/delete `/oem/93_spdk.yaml`, depending on whether the V2 Data Engine is enabled, and whether or not a given node has the `node.longhorn.io/disable-v2-data-engine` label set to `true`.
+
+To implement all the above, we can:
+- In harvester:
+    - Register a syncer for the `longhorn-v2-data-engine` setting. When the value of the setting changes, we do two things:
+        1. Update the NodeConfig for each node to indicate that the V2 Data Engine should be enabled/disabled, taking into account the value of the node's `node.longhorn.io/disable-v2-data-engine` label.  This change will then be picked up by node-manager, much as is currently done for NTP config.
+        2. Write the setting to Longhorn's `lhs/v2-data-engine` setting.
+    - Watch for changes to node labels.  If `node.longhorn.io/disable-v2-data-engine` changes, again, update the NodeConfig for each node appropriately.
+- In node-manager:
+    - When a given NodeConfig changes to enable the V2 engine:
+        - Create `/oem/93_spdk.yaml`
+        - Run `modprobe vfio_pci ; modprobe uio_pci_generic ; modprobe nvme_tcp`
+        - Run `echo 1024 > /sys/kernel/mm/hugepages/hugepages-2048kB/nr_hugepages`
+        - If any of the modprobes fail, or the hugepages can't be allocated, record this somewhere so that it can be picked up in the Harvester GUI with a suggestion to the user to reboot the node (TODO: figure out how to actually do this).
```

Comment:
We could also add the event for a record of whether this procedure is successful or not.

If we enable it on the UI, we could pop a warning/error with the validator.
We could update the system when enabling the v2 engine on UI.
Then, the validator could try these commands and return the error or warning we wanted.